ops-catalog/notify-failure-logs.sql

```diff
@@ -0,0 +1,223 @@
+INSERT INTO failures (shard, ts)
+    SELECT
+        $name,
+        $ts
+    WHERE $message = 'shard failed';
+
+INSERT INTO related_log_lines (shard, ts, level, message)
+    SELECT
+        $name,
+        $ts,
+        $level,
+        $message
+    WHERE $level in ('error', 'warn') AND $message != 'shard failed'
+    ON CONFLICT DO NOTHING;
+
+-- Only keep around the last 10 log lines
+DELETE FROM related_log_lines
+WHERE (shard, ts) IN
+    (
+        SELECT shard, ts from related_log_lines
+        WHERE shard = $name
+        ORDER BY ts DESC
+        LIMIT -1 OFFSET 10
+    );
+
+WITH events_since_previous_notification as (
```

Comment:
Not requesting changes, given this is already deployed and clearly works, but this is hard to follow and I'm curious for your thoughts on how it could be improved for readability.

One pattern I've been playing with myself is a `globals` table that holds any and all global-scope variables. I'll make its columns nullable and then have a single `INSERT INTO globals DEFAULT VALUES;` in the migration. Lambdas then `UPDATE` its one row as needed.

Pragmatically, I've found that it reduces my use of CTEs and lets me express logic a bit more linearly, in smaller chunks. Another benefit is that you can define common views which work off of `globals`, such as building up a complex output, and then trivially re-use them across multiple lambdas.

Going further down the rabbit hole: you can also model functions via a table like `create table do_the_thing(arg1 text, arg2 boolean)`, and then define an after-insert trigger which contains one or more statements, including a statement that deletes the inserted `rowid`. Multiple triggers can be defined, each having a `WHEN` clause, which allows for modeling conditionals. Triggers can also be called recursively, up to a limit.

Clearly this ☝️ can get a bit nuts, but it was nice to discover that it's not _too_ crazy to model a complex, procedural, branching control flow using only SQLite.

Reply:
Yeah, I thought this would be more interesting/useful than it was. I ended up not sending it to slack but forgot to remove the calculation for it. 

Wrt the procedural, branching control flow in SQL(ite), yeah it took me a second to wrap my head around how to do what I wanted. For example, I needed to build up an array of objects to send to Slack and conditionally include certain fields/objects inside it. What I ended up with was this stuff:

```sql
        select JSON_GROUP_ARRAY(value)
        from JSON_EACH(
            JSON_ARRAY(
                CASE (select count(*) from json_each(return_values.log_lines))
                    WHEN 0 THEN NULL
                    ELSE JSON_OBJECT('type', 'divider')
                END,
                CASE (select count(*) from json_each(return_values.log_lines))
                    WHEN 0 THEN NULL
                    ELSE JSON_OBJECT(
                        'type','section',
                        'text',JSON_OBJECT(
                            'type', 'mrkdwn',
                            'text', '*Adjacent Log Messages:*'
                        )
                    )
                END
            )
        )
        WHERE type != 'null'
```